src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs

```diff
@@ -51,12 +48,17 @@ public CoraxIndexReadOperation(Index index, Logger logger, Transaction readTrans
             if (isDistinctCount)
                 pageSize = int.MaxValue;
             var position = query.Start;
-
-            IQueryMatch result = _coraxQueryEvaluator.Search(query, fieldsToFetch);
+            var take = Math.Min(pageSize, Convert.ToInt32(_indexSearcher.CountItemsInIndex)) + position;
+            IQueryMatch result = _coraxQueryEvaluator.Search(query, fieldsToFetch, take);
             if (result == null)
                 yield break;
-
-            var ids = ArrayPool<long>.Shared.Rent(BufferSize);
+            
+            int bufferSize = BufferSize;
+            if (result is SortingMatch or SortingMultiMatch)
+                bufferSize = Convert.ToInt32(_indexSearcher.CountItemsInIndex);
```

Comment:
That might create very large allocation. I wonder if we should have some limit here for using `ArrayPool`.

@ayende @redknightlois any advice on this? @maciejaszyk created the following issue to introduce some cache - https://issues.hibernatingrhinos.com/issue/RavenDB-17631